automerge-backend/src/backend.rs

```diff
@@ -11,26 +12,118 @@ use crate::{
     actor_map::ActorMap,
     change::encode_document,
     error::AutomergeError,
-    event_handlers::{EventHandlerId, EventHandlers},
+    event_handlers::{
+        EventHandler, EventHandlerId, EventHandlers, SendableEventHandler, UnsendableEventHandler,
+    },
     op_handle::OpHandle,
     op_set::OpSet,
     patches::{generate_from_scratch_diff, IncrementalPatch},
-    Change, EventHandler,
+    Change,
 };
 
+/// An automerge Backend.
+///
+/// This struct does not implement Send, if you need Send functionality switch to a
+/// [`SendableBackend`] instead.
 #[derive(Debug, Default, Clone)]
-pub struct Backend {
+pub struct Backend(GenericBackend<UnsendableEventHandler>);
+
+impl Backend {
+    pub fn new() -> Self {
+        Self(GenericBackend::new())
+    }
+
+    pub fn load(data: Vec<u8>) -> Result<Self, AutomergeError> {
+        Ok(Self(GenericBackend::load(data)?))
+    }
+}
+
+impl Deref for Backend {
+    type Target = GenericBackend<UnsendableEventHandler>;
+
+    fn deref(&self) -> &<Self as Deref>::Target {
+        &self.0
+    }
+}
+
+impl DerefMut for Backend {
+    fn deref_mut(&mut self) -> &mut Self::Target {
+        &mut self.0
+    }
+}
+
+/// An automerge Backend that implements Send.
+#[derive(Debug, Default, Clone)]
+pub struct SendableBackend(GenericBackend<SendableEventHandler>);
+
+impl SendableBackend {
+    pub fn new() -> Self {
+        Self(GenericBackend::new())
+    }
+
+    pub fn load(data: Vec<u8>) -> Result<Self, AutomergeError> {
+        Ok(Self(GenericBackend::load(data)?))
+    }
+}
+
+impl Deref for SendableBackend {
+    type Target = GenericBackend<SendableEventHandler>;
+
+    fn deref(&self) -> &<Self as Deref>::Target {
+        &self.0
+    }
+}
+
+impl DerefMut for SendableBackend {
+    fn deref_mut(&mut self) -> &mut Self::Target {
+        &mut self.0
+    }
+}
+
+/// An automerge backend that is generic over the event handler it uses.
+///
+/// This is to enable both sendable and non-sendable backends to exist.
+#[derive(Debug)]
+pub struct GenericBackend<H: EventHandler> {
     queue: Vec<Change>,
     op_set: OpSet,
     states: HashMap<amp::ActorId, Vec<usize>>,
     actors: ActorMap,
     history: Vec<Change>,
     history_index: HashMap<amp::ChangeHash, usize>,
-    event_handlers: EventHandlers,
+    event_handlers: EventHandlers<H>,
 }
 
-impl Backend {
-    pub fn new() -> Self {
+impl<H: EventHandler> Default for GenericBackend<H> {
+    fn default() -> Self {
+        Self {
+            queue: Vec::default(),
+            op_set: OpSet::default(),
+            states: HashMap::default(),
+            actors: ActorMap::default(),
+            history: Vec::default(),
+            history_index: HashMap::default(),
+            event_handlers: EventHandlers::default(),
+        }
+    }
+}
+
+impl<H: EventHandler> Clone for GenericBackend<H> {
```

Comment:
Why do we need a hand rolled `Clone`?

Reply:
It was because otherwise the `EventHandlers` clone doesn't clone but instead made an empty vec, I don't think this PR is needed if we remove queueing in the backend (or at least have the option to).